python/ray/rllib/ppo/loss.py

```diff
@@ -39,17 +40,49 @@ def __init__(
             self.value_function = tf.reshape(self.value_function, [-1])
 
         # Make loss functions.
-        self.ratio = tf.exp(self.curr_dist.logp(actions) -
-                            self.prev_dist.logp(actions))
-        self.kl = self.prev_dist.kl(self.curr_dist)
-        self.mean_kl = tf.reduce_mean(self.kl)
-        self.entropy = self.curr_dist.entropy()
-        self.mean_entropy = tf.reduce_mean(self.entropy)
-        self.surr1 = self.ratio * advantages
-        self.surr2 = tf.clip_by_value(self.ratio, 1 - config["clip_param"],
-                                      1 + config["clip_param"]) * advantages
-        self.surr = tf.minimum(self.surr1, self.surr2)
-        self.mean_policy_loss = tf.reduce_mean(-self.surr)
+
+        curr_logp = self.curr_dist.logp(actions)
+        if isinstance(curr_logp, list):
+            prev_logp = self.prev_dist.logp(actions)
+            self.ratio = np.asarray([tf.exp(a - b) for a, b
+                                     in zip(curr_logp, prev_logp)])
+            self.kl = self.prev_dist.kl(self.curr_dist)
+            self.mean_kl = [tf.reduce_mean(mean_kl_i) for mean_kl_i in self.kl]
+            self.entropy = self.curr_dist.entropy()
+            self.mean_entropy = [tf.reduce_mean(entropy_i)
+                                 for entropy_i in self.entropy]
+            self.surr1 = [ratio_i * advantages for ratio_i in self.ratio]
+            self.surr2 = [advantages *
+                          tf.clip_by_value(ratio_i,
+                                           1 - config["clip_param"],
+                                           1 + config["clip_param"])
+                          for ratio_i in self.ratio]
+            self.surr = [tf.minimum(surr1_i, surr2_i) for surr1_i, surr2_i
+                         in zip(self.surr1, self.surr2)]
+            self.mean_policy_loss = tf.reduce_mean(-tf.add_n(self.surr))
+            self.surr = tf.add_n(self.surr)
+            self.entropy = tf.add_n(self.entropy)
+
+        else:
+            self.ratio = tf.exp(self.curr_dist.logp(actions) -
```

Comment:
Is it possible to handle this as the single-list case, to avoid duplicating the code above?
e.g.
```
if not isinstance(curr_logp, list):
   curr_logp = [curr_logp]
   etc
```

Reply:
Excellent point.